shard.go

```diff
@@ -0,0 +1,97 @@
+package wavelet
+
+import (
+	"sync"
+)
+
+// TransactionMap is a sharded map with TransactionID as its key
+type TransactionMap map[uint64]*transactionMapShard
+
+type transactionMapShard struct {
+	items map[TransactionID]interface{}
+	lock  *sync.RWMutex
+}
+
+const (
+	shardsCount = 64
+)
+
+func NewTransactionMap() TransactionMap {
+	shards := make(TransactionMap, shardsCount)
+
+	for i := uint64(0); i < shardsCount; i++ {
+		shards[i] = &transactionMapShard{
+			items: make(map[TransactionID]interface{}, 2048),
+			lock:  new(sync.RWMutex),
+		}
+	}
+
+	return shards
+}
+
+func (c TransactionMap) Len() uint64 {
+	// Lock all shards
+	for i := uint64(0); i < shardsCount; i++ {
+		c[i].lock.RLock()
+		defer c[i].lock.RUnlock()
+	}
+
+	var l uint64
+	for i := uint64(0); i < shardsCount; i++ {
+		l += uint64(len(c[i].items))
```

Comment:
I think this could be locked, incremented, then unlocked afterwards right in one loop.